hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/oauth2/WorkloadIdentityTokenProvider.java

```diff
@@ -0,0 +1,108 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.fs.azurebfs.oauth2;
+
+import java.io.File;
+import java.io.IOException;
+
+import org.apache.hadoop.thirdparty.com.google.common.base.Strings;
+import org.apache.hadoop.util.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.commons.io.FileUtils;
+
+
+/**
+ * Provides tokens based on Azure AD Workload Identity.
+ */
+public class WorkloadIdentityTokenProvider extends AccessTokenProvider {
+
+  private static final String OAUTH2_TOKEN_PATH = "/oauth2/v2.0/token";
+  private final String authEndpoint;
+
+  private final String clientId;
+
+  private final String tokenFile;
+
+  private long tokenFetchTime = -1;
+
+  private static final long ONE_HOUR = 3600 * 1000;
+
+  private static final Logger LOG = LoggerFactory.getLogger(AccessTokenProvider.class);
+
+  public WorkloadIdentityTokenProvider(final String authority, final String tenantId,
+      final String clientId, final String tokenFile) {
+    Preconditions.checkNotNull(authority, "authority");
+    Preconditions.checkNotNull(tenantId, "tenantId");
+    Preconditions.checkNotNull(clientId, "clientId");
+    Preconditions.checkNotNull(tokenFile, "tokenFile");
+
+    this.authEndpoint = authority + tenantId + OAUTH2_TOKEN_PATH;
+    this.clientId = clientId;
+    this.tokenFile = tokenFile;
+  }
+
+  @Override
+  protected AzureADToken refreshToken() throws IOException {
+    LOG.debug("AADToken: refreshing token from JWT Assertion");
+    String clientAssertion = getClientAssertion(tokenFile);
+    AzureADToken token = AzureADAuthenticator
+        .getTokenUsingJWTAssertion(authEndpoint, clientId, clientAssertion);
+    tokenFetchTime = System.currentTimeMillis();
+    return token;
+  }
+
+  /**
+   * Checks if the token is about to expire as per base expiry logic.
+   * Otherwise try to expire every 1 hour.
+   *
+   * @return true if the token is expiring in next 1 hour or if a token has
+   * never been fetched
+   */
+  @Override
+  protected boolean isTokenAboutToExpire() {
+    if (tokenFetchTime == -1 || super.isTokenAboutToExpire()) {
+      return true;
+    }
+
+    boolean expiring = false;
+    long elapsedTimeSinceLastTokenRefreshInMillis =
+        System.currentTimeMillis() - tokenFetchTime;
+    // In case token is not refreshed for 1 hr or any clock skew issues,
+    // refresh token.
+    expiring = elapsedTimeSinceLastTokenRefreshInMillis >= ONE_HOUR
+        || elapsedTimeSinceLastTokenRefreshInMillis < 0;
+    if (expiring) {
+      LOG.debug("JWTToken: token renewing. Time elapsed since last token fetch:"
+          + " {} milliseconds", elapsedTimeSinceLastTokenRefreshInMillis);
+    }
+
+    return expiring;
+  }
+
+  private static String getClientAssertion(String tokenFile)
```

Comment:
This method doesn't use any member variables so I made it static.  I can remove `static` if you prefer.

Reply:
Lets not have it static, reason being, it is being called only from single non-static method. We can remove the `tokenFile` argument and let it depend on the object field.

Reply:
Also, reading will be done for each refresh call, let call it from the constructor itself. This will be better as: if the file read fails for any reason, it will be raised on the FileSystem creation itself, and also would be saving IO calls.

Reply:
> Lets not have it static, reason being, it is being called only from single non-static method. We can remove the tokenFile argument and let it depend on the object field.

Will do.

> Also, reading will be done for each refresh call, let call it from the constructor itself. This will be better as: if the file read fails for any reason, it will be raised on the FileSystem creation itself, and also would be saving IO calls.

Doing that will introduce a bug where the ABFS driver will successfully authenticate the first time when `getClientAssertion` is called from the constructor, but will fail during subsequent invocations of `getClientAssertion` because the token in the file is refreshed out-of-band by Azure.  Subsequent invocations of `refreshToken` will return the token that was cached when the constructor was invoked instead of reading the new token that was written to `tokenFile` by Azure.

For your reference, this `WorkloadIdentityTokenProvider` class has logic that is similar to the [WorkloadIdentityCredential](https://github.com/Azure/azure-sdk-for-java/blob/0d2d096b5f9f5dfbf512040a6e8345b4cdac2a6e/sdk/identity/azure-identity/src/main/java/com/azure/identity/WorkloadIdentityCredential.java#L56) class in Microsoft's Azure Identity SDK for Java. That class stores the path to the token file ([federatedTokenFilePathInput](https://github.com/Azure/azure-sdk-for-java/blob/0d2d096b5f9f5dfbf512040a6e8345b4cdac2a6e/sdk/identity/azure-identity/src/main/java/com/azure/identity/WorkloadIdentityCredential.java#L80)) in the constructor and [reads from the token file](https://github.com/Azure/azure-sdk-for-java/blob/0d2d096b5f9f5dfbf512040a6e8345b4cdac2a6e/sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java#L185) every time a new token is requested.

Reply:
If I am understanding correctly, there is some process on the azure vm which refreshes this file? Would be great if you can please cite the reference and also if we could add as a javadoc for the method `getClientAssertion`.
As, it is important that the tokenFile is not cached, lets add a test around this as well.

Reply:
>there is some process on the azure vm which refreshes this file?

Yes.

>please cite the reference

I added the [citation](https://azure.github.io/azure-workload-identity/docs/faq.html#does-workload-identity-work-in-disconnected-environments) to the javadoc for `getClientAssertion`.

>As, it is important that the tokenFile is not cached, lets add a test around this as well.

The only way I see to verify the token file is not cached is to somehow mock the call to `AzureADAuthenticator.getTokenUsingJWTAssertion()` in `WorkloadIdentityTokenProvider.refreshToken()`, which is problematic. Please see my comments [here](https://github.com/apache/hadoop/pull/5953/files#r1453634534) regarding my thoughts on mocking it.  If we can identify a way to mock that code then I can add a unit test to verify the token file is not cached. Otherwise, I don't see how to implement that unit test.